### ISSUE TITLE
Runtime error / SIGSEGV on Linux i386 with precompiled binary

### ISSUE THREAD

**Issue by smartmic** [lmorg/murex#322]
Hi there, 
I wanted to try out murex on my old Linux box but the precompiled binary does not run. Any idea or system requirements / minimum Linux kernel version which I am missing?

The system:

    $ uname -a
    Linux crunchbang 3.2.0-6-686-pae #1 SMP Debian 3.2.102-1 i686 GNU/Linux

The output when trying to start murex:

 ```
$ ./murex-linux-386
Loading profile `.murex_preload`
Loading profile `.murex_profile`
~/downloads » panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x804a18b]

goroutine 1 [running]:
runtime/internal/atomic.Store64(0x9ca6304, 0x0, 0x0)
        /go/go/src/runtime/internal/atomic/asm_386.s:205 +0xb
github.com/lmorg/murex/utils/readline.(*Instance).Readline(0x9ca62c0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/lmorg/murex/utils/readline/readline.go:47 +0x25a
github.com/lmorg/murex/shell.ShowPrompt()
        /go/src/github.com/lmorg/murex/shell/shell.go:119 +0x178
github.com/lmorg/murex/shell.Start()
        /go/src/github.com/lmorg/murex/shell/shell.go:70 +0x195
main.startMurex()
        /go/src/github.com/lmorg/murex/main.go:153 +0x1ab
main.main()
        /go/src/github.com/lmorg/murex/main.go:36 +0xb4
```

**Comment by lmorg** [comment]
@smartmic 

Thanks for reporting that. I'll take a look this weekend.

**Comment by lmorg** [comment]
@smartmic 

Update: I haven't yet been able to spin up a 32bit version of Linux to replicate your error (largely because my PXE server keeps pushing 64bit images even when I fire up a 32bit VM -- so I've got to fix that first!) however I have done a little bit of digging into the code around that's reported in the stack trace and it might be a compiler / corelib bug. The Go version used to compile those binaries is a few versions out of date and the code around `asm_386.s` appears to have been completely rewritten.

Given I'm unlikely to get time to do some proper comparison testing between Go versions this week, I'm tempted to just re-run the CI/CD pipeline with the latest version of Go and see if that spits out better binaries....if you're happy to download and test that yourself?

I'll push updated build scripts shortly so ~30 mins for updated executables (look for version `3.1.4000` or above).

**Comment by smartmic** [comment]
@lmorg shure, I would like to test those binaries. Where can I find them? Since my 32bit Linux is quite limited in resources, I have neither Go or Docker installed. I hope this is not needed on my end.

**Comment by lmorg** [comment]
@smartmic 

The latest binaries can be downloaded from https://murex.rocks/DOWNLOAD.html and they should now be built with the latest version of Go. So you can just `curl` / `wget`. Though you will need to `gunzip` the file before it will execute.

**Comment by smartmic** [comment]
Okay, that is what I did. I was wondering because the `--version` argument spits out a different version.  At least, this option in the runtime works. But I still could not  start the murex shell, same error as above. 

In detail:
```bash
~/bin$ wget https://murex.rocks/bin/latest/murex-linux-386.gz
--2021-05-09 21:05:37--  https://murex.rocks/bin/latest/murex-linux-386.gz
Resolving murex.rocks (murex.rocks)... 13.226.155.63, 13.226.155.90, 13.226.155.5, ...
Connecting to murex.rocks (murex.rocks)|13.226.155.63|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6674135 (6.4M) [application/x-gzip]
Saving to: `murex-linux-386.gz'

100%[===================================================================================================================>] 6,674,135   10.7M/s   in 0.6s

2021-05-09 21:05:40 (10.7 MB/s) - `murex-linux-386.gz' saved [6674135/6674135]

~/bin$ gunzip murex-linux-386.gz
gzip: murex-linux-386 already exists; do you wish to overwrite (y or n)? y
~/bin$ chmod +x murex-linux-386
~/bin$ ./murex-linux-386 --version
murex v2.1.3200 BETA
Licence GPL v2
© 2018-2021 Laurence Morgan
~/bin$ ./murex-linux-386
Loading profile `.murex_preload`
Loading profile `.murex_profile`
~/bin » panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x804a18b]

goroutine 1 [running]:
runtime/internal/atomic.Store64(0x98a8304, 0x0, 0x0)
        /go/go/src/runtime/internal/atomic/asm_386.s:205 +0xb
github.com/lmorg/murex/utils/readline.(*Instance).Readline(0x98a82c0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/lmorg/murex/utils/readline/readline.go:47 +0x25a
github.com/lmorg/murex/shell.ShowPrompt()
        /go/src/github.com/lmorg/murex/shell/shell.go:119 +0x178
github.com/lmorg/murex/shell.Start()
        /go/src/github.com/lmorg/murex/shell/shell.go:70 +0x195
main.startMurex()
        /go/src/github.com/lmorg/murex/main.go:153 +0x1ab
main.main()
        /go/src/github.com/lmorg/murex/main.go:36 +0xb4
```

**Comment by lmorg** [comment]
@smartmic 

I'm really glad you posted the `--version` information because that highlighted that the CI/CD pipeline mysteriously stopped pushing the executables while not failing the pipeline. All very odd.

Anyhow, I've fixed the CI/CD pipeline so, if you don't mind, can you try again.

Also apologies for the inconvenience there. Literally the one thing I don't check when pushing a new version is the executables because it's the same CI/CD pipeline as the website and if the website version changes (which it did do) and the pipeline hadn't failed then the executables _should_ have also updated. Clearly that wasn't the case here. But I've manually checked and the files are, _now_, up to date.

**Comment by lmorg** [comment]
Hi, @smartmic 

I doubt compiling it on your arch would make any difference. Looking through the code, readline is using a 64bit atomic operation whereas all the other atomic operations are 32bit. I can't see anything in any of the Go language docs to say there is an issue with using 64bit atomic operations on 32bit operating systems but the fact remains we're getting errors using them where we're not on the other 32bit operations. So I've changed those 64bit operations to 32bit (IIRC it's a date stamp and 64bit was just for futureproofing, but there will be plenty of updates to _murex_ between now and 2038 :) )

Unfortunately I'm unable to test this because VirtualBox is refusing to install 32bit x86 distros on my dev machine. So I've just merged the code in the expectation it doesn't break things further. :D 

If you're happy to do more testing, it's version `2.1.4020` or greater that you're looking for. I've literally only just merged the code but hopefully by the time you read this the CI/CD pipeline would have finished and the latest binaries will be up.
